pageserver/src/tenant_mgr.rs

```diff
@@ -321,12 +310,88 @@ pub async fn detach_tenant(
     // we will attempt to remove files which no longer exist. This can be fixed by having shutdown
     // mechanism for tenant that will clean temporary data to avoid any references to ephemeral files
     let local_tenant_directory = conf.tenant_path(&tenant_id);
-    fs::remove_dir_all(&local_tenant_directory).with_context(|| {
-        format!(
-            "Failed to remove local tenant directory '{}'",
-            local_tenant_directory.display()
-        )
-    })?;
+    tokio::fs::remove_dir_all(&local_tenant_directory)
+        .await
+        .with_context(|| {
+            format!("Failed to remove local tenant directory {local_tenant_directory:?}")
+        })?;
+
+    Ok(())
+}
+
+pub async fn schedule_tenant_load(
+    conf: &'static PageServerConf,
+    tenant_id: TenantId,
+    remote_storage: Option<GenericRemoteStorage>,
+) -> anyhow::Result<()> {
+    let tenant_path = conf.tenant_path(&tenant_id);
+    anyhow::ensure!(
+        tenant_path.is_dir(),
+        "Cannot load tenant from path {tenant_path:?}, it either does not exist or not a directory"
+    );
+    anyhow::ensure!(
+        !crate::is_temporary(&tenant_path),
+        "Cannot load tenant from temporary path {tenant_path:?}"
+    );
+    anyhow::ensure!(
+        !tenant_path.is_empty_dir().with_context(|| {
+            format!("Failed to check whether {tenant_path:?} is an empty dir")
+        })?,
+        "Cannot load tenant from empty directory {tenant_path:?}"
+    );
+    let tenant_attach_mark = conf.tenant_attaching_mark_file_path(&tenant_id);
+    anyhow::ensure!(
+        !tenant_attach_mark.exists(),
```

Comment:
This will break the use case `/ignore` & `/spawn_load` a tenant that has encountered an error while it was attaching.

In general, I think most of the checks in this function before `run_if_no_tenant_in_memory` are kind of redundant. `load_local_tenant` and the `spawn_load` / `spawn_attach` functions already do them.

Reply:
Right

> most of the checks in this function before `run_if_no_tenant_in_memory` are kind of redundant

Start to agree more when I've thought that it's better to put them after we take the lock: more chances to prevent a race on concurrent `attach`/`load` requests scheduled.

So, now we check the in-memory state, grab the lock, and then, inside `load_local_tenant` do more checks to ensure that this sort of public method does not load some ignored tenant by accident.
This way, a few FS checks are duplicated still, but that seems not much.

I've removed the attached file check and added a test for such scenario.
Looks like another "fill the DB and upload layers" common function slowly emerges in Python tests, but I've not moved it into the lib code yet.